src/filter.rs

```diff
@@ -0,0 +1,297 @@
+                buf.as_ptr() as *const c_char, buf.len())
+        };
+        call::try(code).and(Ok(()))
+    }
+
+    fn close(&mut self) -> Result<(), Error> {
+        let code = unsafe { ((*self.stream).close)(self.stream) };
+        call::try(code).and(Ok(()))
+    }
+}
+
+extern fn stream_write(stream: *mut git_writestream,
+                       buf: *const c_char,
+                       len: size_t) -> c_int {
+    let ws = stream as *mut Stream;
+    unsafe {
+        let input = slice::from_raw_parts(buf as *const u8, len);
+        match (*ws).handler.write(input) {
+            Ok(..) => raw::GIT_OK,
+            Err(_) => raw::GIT_ERROR,
+        }
+    }
+}
+
+extern fn stream_free(stream: *mut git_writestream) {
+    let _ = unsafe { Box::from_raw(stream as *mut Stream) };
+}
+
+extern fn stream_close(stream: *mut git_writestream) -> c_int {
+    let ws = stream as *mut Stream;
+    unsafe {
+        match (*ws).handler.close() {
+            Ok(..) => raw::GIT_OK,
+            Err(_) => raw::GIT_ERROR,
+        }
+    }
+}
+
+extern fn filter_init(_filter: *mut git_filter) -> c_int {
+    raw::GIT_OK
+}
+
+extern fn filter_shutdown(filter: *mut git_filter) {
+    let _ = unsafe { Box::from_raw(filter as *mut Filter) };
+    // dropped
+}
+
+extern fn filter_stream_init(out: *mut *mut raw::git_writestream,
+                        filter: *mut git_filter,
+                        _payload: *mut *mut c_void,
+                        src: *const raw::git_filter_source,
+                        next: *mut raw::git_writestream) -> c_int
+{
+    let filter = filter as *mut Filter;
+    let next = RawWriteStream { stream: next };
+    let source = unsafe { FilterSource::from_raw(src) };
+    let stream = unsafe { match (*filter).instance.open_stream(source, next) {
+        Ok(s) => s,
+        Err(_) => return raw::GIT_ERROR,
+    }};
+    let mut stream = Box::new(Stream {
+        parent: git_writestream {
+            write: stream_write,
+            close: stream_close,
+            free: stream_free,
+        },
+        handler: stream,
+    });
+    unsafe {
+        let p = (&mut *stream) as *mut Stream as *mut git_writestream;
+        *out = p;
+    }
+    mem::forget(stream);
+    raw::GIT_OK
+}
+
+impl Filter {
+    /// create a command filter
+    fn new<T: StreamFilter>(user_filter: T) -> Filter {
+        let f = Filter {
+            filter: git_filter {
+                version: raw::GIT_FILTER_VERSION,
+                attributes: ptr::null(),
+                initialize: Some(filter_init),
+                shutdown: Some(filter_shutdown),
+                // this callback is not needed, custom filter can implement
+                // passthrough easily
+                check: None,
+
+                // do not use legency filter model
+                apply: None,
+                stream: Some(filter_stream_init),
+
+                // stream filter do not need this
+                cleanup: None,
```

Comment:
Hm is `check` important as an optimization to avoid unnecessarily creating streams?

Thanks for the info about `apply`! Could you expand the comment here to say that it's an essentially deprecated way of doing things in libgit2?

Also, should `cleanup` get bound here?